### ISSUE TITLE
Canal Pods no longer starting after upgrading from Kubernetes 1.7.x to 1.8.x 

### ISSUE THREAD

**Issue by awslovato** [projectcalico/canal#108]
After upgrading my cluster my canal pods are failing with Status CrashLoopBackOff on master and all nodes. It could be due to incompatibility introduced in 1.8.

## Expected Behavior

If there is no reason for this to happen all canal pods should be starting and be in Ready state on all nodes and master.

## Steps to Reproduce (for bugs)

1. cordon host to upgrade node to Kubernetes 1.8
2. reboot host
3. when host is up, uncordon the host to get the ready state from the kubelet.
4. Host becomes ready, but canal daemonset does not stay up on any node or master.

## Context

Currently with no CNI deployments either error out or stay terminated.

```
[root@master ~]# kubectl get pods --all-namespaces
NAMESPACE            NAME                                    READY     STATUS             RESTARTS   AGE
kube-system          canal-1c8dr                             2/3       CrashLoopBackOff   48         73d
kube-system          canal-7srsr                             0/3       Completed          32         73d
kube-system          canal-hpj5w                             2/3       CrashLoopBackOff   38         66d
kube-system          canal-vrdxj                             2/3       CrashLoopBackOff   68         73d
kube-system          canal-xpth2                             2/3       CrashLoopBackOff   35         52d
kube-system          etcd-dfsbxvoconm1                       1/1       Running            2          14d
kube-system          kube-apiserver-dfsbxvoconm1             1/1       Running            1          35m
kube-system          kube-controller-manager-dfsbxvoconm1    1/1       Running            1          35m
kube-system          kube-dns-545bc4bfd4-qz25d               3/3       Running            3          23m
kube-system          kube-proxy-5fplb                        1/1       Running            1          32m
kube-system          kube-proxy-6fdqt                        0/1       Error              0          34m
kube-system          kube-proxy-qxhrp                        1/1       Running            1          32m
kube-system          kube-proxy-vcvhb                        1/1       Running            1          33m
kube-system          kube-proxy-vmrbq                        1/1       Running            1          34m
kube-system          kube-scheduler-dfsbxvoconm1             1/1       Running            1          35m
kube-system          kubernetes-dashboard-3313488171-twfq2   1/1       Running            0          9m
```
```
[root@master ~]# kubectl logs canal-1c8dr --container calico-node --namespace=kube-system
time="2017-10-16T23:13:30Z" level=info msg="NODENAME environment not specified - check HOSTNAME"
time="2017-10-16T23:13:30Z" level=info msg="Loading config from environment"
Checking datastore connection
Datastore connection verified
time="2017-10-16T23:13:30Z" level=info msg="Datastore is ready"
IPv4 address $NODE_IP discovered on interface ens160
No AS number configured on node resource, using global value
time="2017-10-16T23:13:30Z" level=info msg="Ensuring ThirdPartyResources exist"
time="2017-10-16T23:13:30Z" level=info msg="Ensuring GlobalConfig ThirdPartyResource exists"
ERROR: Unable to set node resource configuration: Failed to ensure ThirdPartyResources exist: resource does not exist: {{ } {global-config.projectcalico.org  kube-system    %!s(int64=0) 0001-01-01 00:00:00 +0000 UTC <nil> %!s(*int64=<nil>) map[] map[] [] [] } Calico Global Configuration [{v1}]}
Terminating
Calico node failed to start
```

The above message is the same for all instances of the canal daemonset, just change the $NODE_IP

If you see anything glaringly off let me know, appreciated!

## Your Environment

* Calico version: node: 1.3.0 cni: 1.9.1
* Flannel version: 0.8.0
* Orchestrator version: v.1.8.1
* Operating System and version: Centos 7.4.1708 Core
* Link to your project (optional):

**Comment by caseydavenport** [comment]
> ERROR: Unable to set node resource configuration: Failed to ensure ThirdPartyResources exist: resource does not exist: {{ } {global-config.projectcalico.org kube-system %!s(int64=0) 0001-01-01 00:00:00 +0000 UTC %!s(*int64=) map[] map[] [] [] } Calico Global Configuration [{v1}]}

Yep, this is expected when running older versions of Canal against kubernetes 1.8


k8s 1.8 removed support for TPRs, which earlier versions of canal relied on. Instead, CRDs are used. You'll need to follow the migration steps provided here to upgrade to the latest canal (it says v1.7, but it's actually v1.7+) https://github.com/projectcalico/canal/tree/master/k8s-install#for-kubernetes-17

As part of that, it will have you run through a migration process to translate the TPR data used by the old Calico version into CRDs supported by the new Calico version.

**Comment by caseydavenport** [comment]
@awslovato did you get a chance to try to upgrade steps listed?

**Comment by awslovato** [comment]
@caseydavenport I did.  The pods created in step 4 to do the upgrade are looking for the data that is supposed to be created in steps 1 -3. I did not have any data there and the logs on the pods show that it's expecting something and crash loop till they hit the limit for instances.  Step 4.2 (which maps to step 5 in the top list?) states:
4.2. Now you can upgrade the calico/node and calico/cni images in your Kubernetes Calico DaemonSet. (Make sure you reboot your calico-node pods one at a time if calico-node updateStrategy is not set to RollingUpdate.)

Since this is supposed to be the step to describe how to do this I'm unsure how the actual upgrade is supposed to proceed as it's not stated.

**Comment by caseydavenport** [comment]
Ok, sounds like a potential upgrade bug.

Could you share the full logs from the upgrade?

**Comment by awslovato** [comment]
Sure can, tell me which pods or services you would like to see.

**Comment by caseydavenport** [comment]
I'm mostly interested in the logs from the upgrade job to see exactly where it's failing. There should be a pod with the name ` calico-upgrade-v2.5-xxxxxx` or similar.

You may need to use `kubectl get pods --show-all` in order to see it. 

It would be useful to see the output of the following commands as well I think:

### Prior to upgrade

```
kubectl get thirdpartyresource
kubectl get ippools
kubectl get globalconfig
```

### After upgrade

```
kubectl get customresourcedefinitions
kubectl get ippools
kubectl get globalfelixconfig
```

**Comment by awslovato** [comment]
[root@dfsbxvoconm1 calico]# kubectl get thirdpartyresource
the server doesn't have a resource type "thirdpartyresource"
[root@dfsbxvoconm1 calico]# kubectl get ippools
No resources found.
[root@dfsbxvoconm1 calico]# kubectl get globalconfig
the server doesn't have a resource type "globalconfig"

Also, none of the pods that failed seemed to generate any logs:
[root@dfsbxvoconm1 ~]# kubectl get pods --show-all
NAME                        READY     STATUS    RESTARTS   AGE
calico-upgrade-v2.5-7c5gg   0/1       Error     0          10d
calico-upgrade-v2.5-8q2b9   0/1       Error     0          10d
calico-upgrade-v2.5-jzvj2   0/1       Error     0          10d
calico-upgrade-v2.5-kvgxm   0/1       Error     0          10d
calico-upgrade-v2.5-ph7h5   0/1       Error     0          10d
calico-upgrade-v2.5-qrdvv   0/1       Error     0          10d
calico-upgrade-v2.5-s6ttr   0/1       Error     0          10d
[root@dfsbxvoconm1 calico]# kubectl logs calico-upgrade-v2.5-7c5gg
[root@dfsbxvoconm1 calico]# kubectl logs calico-upgrade-v2.5-8q2b9
[root@dfsbxvoconm1 calico]# kubectl logs calico-upgrade-v2.5-jzvj2
[root@dfsbxvoconm1 calico]# kubectl logs job/calico-upgrade-v2.5
Found 7 pods, using pod/calico-upgrade-v2.5-jzvj2
[root@dfsbxvoconm1 calico]#
(etc.)

[root@dfsbxvoconm1 calico]# kubectl describe clusterroles calico-25-migration
Name:         calico-25-migration
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"rbac.authorization.k8s.io/v1beta1","kind":"ClusterRole","metadata":{"annotations":{},"name":"calico-25-migration","namespace":""},"rules...
PolicyRule:
  Resources                                       Non-Resource URLs  Resource Names  Verbs
  ---------                                       -----------------  --------------  -----
  bgppeers.crd.projectcalico.org                  []                 []              [create get list post patch update watch]
  customresourcedefinitions.apiextensions.k8s.io  []                 []              [create get list post patch update watch]
  globalbgpconfigs.crd.projectcalico.org          []                 []              [create get list post patch update watch]
  globalbgpconfigs.projectcalico.org              []                 []              [create get list post patch update watch]
  globalbgppeers.projectcalico.org                []                 []              [create get list post patch update watch]
  globalconfigs.projectcalico.org                 []                 []              [create get list post patch update watch]
  globalfelixconfigs.crd.projectcalico.org        []                 []              [create get list post patch update watch]
  globalnetworkpolicies.crd.projectcalico.org     []                 []              [create get list post patch update watch]
  ippools.crd.projectcalico.org                   []                 []              [create get list post patch update watch]
  ippools.projectcalico.org                       []                 []              [create get list post patch update watch]
  nodes                                           []                 []              [create get list post patch update watch]
  thirdpartyresources.extensions                  []                 []              [create get list post patch update watch]
[root@dfsbxvoconm1 calico]# kubectl describe clusterrolebinding calico-25-migration
Name:         calico-25-migration
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"rbac.authorization.k8s.io/v1beta1","kind":"ClusterRoleBinding","metadata":{"annotations":{},"name":"calico-25-migration","namespace":""}...
Role:
  Kind:  ClusterRole
  Name:  calico-25-migration
Subjects:
  Kind            Name                 Namespace
  ----            ----                 ---------
  ServiceAccount  calico-25-migration  default

**Comment by awslovato** [comment]
Should I just destroy the existing daemon set and recreate it with the 1.7 manifests?

**Comment by caseydavenport** [comment]
```
[root@dfsbxvoconm1 calico]# kubectl get thirdpartyresource
the server doesn't have a resource type "thirdpartyresource"
```

That seems to indicate you're running a Kubernetes v1.8 cluster, which could be the problem - the migration scripts are meant to run on a Kubernetes v1.7 cluster.

> Should I just destroy the existing daemon set and recreate it with the 1.7 manifests?

That might work - you would lose any configured `globalconfig` or `ippools` but it looks like you might not have any of those anyway.

**Comment by awslovato** [comment]
> That seems to indicate you're running a Kubernetes v1.8 cluster, which could be the problem - the migration scripts are meant to run on a Kubernetes v1.7 cluster.
I have that as step 1 in my problem description.

What do you recommend for a 1.8 cluster?

**Comment by caseydavenport** [comment]
@awslovato yeah, I noticed that and I realize now I didn't make it super clear that the upgrade instructions I linked to need to be run on a Kubernetes v1.7 cluster prior to upgrading to Kubernetes v1.8.

The steps are roughly:

- Upgrade to k8s v1.7 (if not already on v1.7)
- Run the migration steps linked above
- upgrade Calico
- Upgrade to k8s v1.8.

[These manifests](https://github.com/projectcalico/canal/tree/master/k8s-install/1.7) are the correct ones for a Kubernetes v1.8 cluster.
